Neo4j.Driver/Neo4j.Driver/Internal/ExceptionHandling/ClientErrorExceptionFactory.cs

```diff
@@ -0,0 +1,120 @@
+        BuildExceptionFactories(codesAndExceptions);
+    }
+
+    private List<(string code, Type exceptionType)> GetCodesAndExceptions()
+    {
+        var exceptionTypes = GetAllNeo4jExceptions();
+
+        return exceptionTypes
+            .Select(
+                exceptionType => new
+                {
+                    exceptionType,
+                    attr = exceptionType.GetCustomAttribute<ClientErrorCodeAttribute>()
+                })
+            .Where(t => t.attr is not null)
+            .Select(t => (t.attr.Code, t.exceptionType))
+            .ToList();
+    }
+
+    private static IEnumerable<Type> GetAllNeo4jExceptions()
+    {
+        var type = typeof(Neo4jException);
+        var assembly = type.Assembly;
+        var types = assembly.GetExportedTypes().Where(t => type.IsAssignableFrom(t));
+        return types;
+    }
+
+    private int CompareByCode((string code, Type exceptionType) x, (string code, Type exceptionType) y)
+    {
+        // x comes before y if y matches x - this would happen if:
+        // x = Error.Specific
+        // y = Error.*
+        // this means that less-specific wildcards are at the end of the list, so the first
+        // matching wildcard will always be the most specific
+
+        if (_simpleWildcardHelper.StringMatches(x.code, y.code))
+        {
+            return -1;
+        }
+
+        if (_simpleWildcardHelper.StringMatches(y.code, x.code))
+        {
+            return 1;
+        }
+
+        return 0;
+    }
+
+    private void BuildExceptionFactories(IEnumerable<(string code, Type exceptionType)> codesAndExceptions)
+    {
+        foreach (var (code, type) in codesAndExceptions)
+        {
+            Func<string, string, Exception, Neo4jException> factory;
+            if (type.GetConstructor(new[] { typeof(string), typeof(string), typeof(Exception) }) is {} threeParamCtr)
+            {
+                factory = (c, m, x) => (Neo4jException)threeParamCtr.Invoke(new object[] { c, m, x });
+            }
+            else if (type.GetConstructor(new[] { typeof(string), typeof(Exception) }) is {} twoParamCtr)
+            {
+                factory = (_, m, x) => (Neo4jException)twoParamCtr.Invoke(new object[] { m, x });
+            }
+            else if (type.GetConstructor(new[] { typeof(string) }) is {} oneParamCtr)
+            {
+                factory = (_, m, _) => (Neo4jException)oneParamCtr.Invoke(new object[] { m });
+            }
+            else
+            {
+                continue;
+            }
+
+            _exceptionFactories.Add(new FactoryInfo(code, factory));
+        }
+    }
+
+    public Neo4jException GetException(string code, string message, Exception innerException = null)
+    {
+        var factoryInfo = _exceptionFactories.FirstOrDefault(f => _simpleWildcardHelper.StringMatches(code, f.Code));
+        var exception = factoryInfo is null
+            ? new Neo4jException(code, message, innerException)
```

Comment:
- This should raise a protocol exception When we call this with non-client error code as this would result in a bug.
- the return type should be ClientException too to reinforce this is only to return client errors or exceptions that inherient ClientException.